### ISSUE TITLE
.DS_Store errors on macOS

### ISSUE THREAD

**Issue by 0rphee** [juhp/stack-clean-old#18]
Hi!
While trying to use this tool I constantly found this errors:

`stack-clean-old: .DS_Store: changeWorkingDirectory: inappropriate type (Not a directory)`
`stack-clean-old: ghc-9.4.6/lib/ghc-9.4.6/lib: removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removeDirectory: unsatisfied constraints (Directory not empty)`

I constantly had to remove the files recursively, however, the system is very quick to create them again, and `stan-clean-old` would fail. Would it be possible to add functionality to this tool to automatically remove this files too?

Thank you for the tool btw!

**Comment by juhp** [comment]
Ahh I see, that is annoying: I should be able to avoid that hopefully, I will have to see..

**Comment by juhp** [comment]
Is it possible you could share a part of a project/folder directory tree exhibiting the problem?
I want to make sure I understand the situation fully.

Say from `tree -a` output.

*edit:* Actually I think `find -name .DS_Store` would be simpler and more informative.

**Comment by juhp** [comment]
Well I guess I can just assume there is a file in every directory...

**Comment by juhp** [comment]
> `stack-clean-old: ghc-9.4.6/lib/ghc-9.4.6/lib: removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removeDirectory: unsatisfied constraints (Directory not empty)`

Is that real output btw? I have never seen such output before.

(Also next time if you could paste the actually commands+output it would be helpful.)

**Comment by juhp** [comment]
Could try this patch and see if it gets rid of all the errors you are seeing?
```patch
diff --git a/src/Directories.hs b/src/Directories.hs
index 62c965a..00f2025 100644
--- a/src/Directories.hs
+++ b/src/Directories.hs
@@ -10,7 +10,7 @@ module Directories (
   )
 where
 
-import Control.Monad (forM_, unless, when)
+import Control.Monad (filterM, forM_, unless, when)
 import Data.List.Extra
 import SimpleCmd ((+-+),
 #if MIN_VERSION_simple_cmd(0,2,0)
@@ -71,7 +71,7 @@ traversePlatforms' getdir msystem act = do
 
 listPlatforms :: Maybe String -> IO [FilePath]
 listPlatforms msystem = do
-  platforms <- listDirectory "."
+  platforms <- listDirectory "." >>= filterM doesDirectoryExist
   case msystem of
     Nothing -> return platforms
     Just s ->
@@ -83,7 +83,7 @@ listPlatforms msystem = do
 
 listCurrentDirectory :: IO [FilePath]
 listCurrentDirectory =
-  filter (\d -> head d /= '.') <$> listDirectory "."
+  filter (\d -> head d /= '.') <$> listDirectory "." >>= filterM doesDirectoryExist
 
 #if !MIN_VERSION_simple_cmd(0,2,0)
 warning :: String -> IO ()
```

**Comment by juhp** [comment]
Could you try #19 and see if it gets rid of the errors you are seeing?

**Comment by 0rphee** [comment]
> Well I guess I can just assume there is a file in every directory...

Yes, that's correct. They are created in every directory automatically. 

> Is that real output btw? I have never seen such output before.
>
> (Also next time if you could paste the actually commands+output it would be helpful.)

Yes, that's real output. I think it was `stack-clean-old purge-older -d` and/or `stack-clean-old remove -d`, outside of a specific project so that it acted on the stack root `~/.stack`. Because the commands didn't work I ended running stuff like this: `find . -name ".DS_Store" -type f -print -delete && stack-clean-old remove 9.4.6 -d`.

Sorry about the commands... forgot about them. 

> Could you try https://github.com/juhp/stack-clean-old/pull/19 and see if it gets rid of the errors you are seeing?

I'll try it shortly.

**Comment by 0rphee** [comment]
I tested creating `.testfile`'s recursively in `~/.stack/snapshots` and a project's `.stack-work` with `find . -type d -exec touch {}/.testfile \;&& find . -name ".testfile" -type f -print`

I tried with the other commands but I couldn't reproduce the notorious `stack-clean-old: ghc-9.4.6/lib/ghc-9.4.6/lib: removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removeDirectory: unsatisfied constraints (Directory not empty)`. Most probably because I accidentally removed `.stack/snapshots/` (whoops!), without many old builds to test the other commands.
```
# old outside and inside project dir
$ stack-clean-old remove 9.2.8 -d
stack ghc compiler version 9.2.8 not found
stack-clean-old: .testfile: changeWorkingDirectory: inappropriate type (Not a directory)

# new outside and inside project dir
$ stack-clean-old remove 9.2.8 -d
stack ghc compiler version 9.2.8 not found
aarch64-osx:
1 dir in aarch64-osx/*/9.2.8 removed

# old outside and inside project dir
$ stack-clean-old keep-minor 9.2 -d
stack-clean-old: .testfile: changeWorkingDirectory: inappropriate type (Not a directory)

# new outside and inside project dir
$ stack-clean-old keep-minor 9.2 -d
```

**Comment by juhp** [comment]
Thanks for testing

> I tried with the other commands but I couldn't reproduce the notorious `stack-clean-old: ghc-9.4.6/lib/ghc-9.4.6/lib: removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removePathRecursive:removeContentsRecursive:removeDirectory: unsatisfied constraints (Directory not empty)`. Most probably because I accidentally removed `.stack/snapshots/` (whoops!), without many old builds to test the other commands.

Oh that's interesting - I can try moving those dirs and see what happens.

> ```
> # new outside and inside project dir
> $ stack-clean-old remove 9.2.8 -d
> stack ghc compiler version 9.2.8 not found
> aarch64-osx:
> 1 dir in aarch64-osx/*/9.2.8 removed
> :
> # new outside and inside project dir
> $ stack-clean-old keep-minor 9.2 -d
> ```

Okay this looks better right?

**Comment by 0rphee** [comment]
Yeah, I think that at least for this cases, the fix has worked :)

**Comment by juhp** [comment]
Great - do let me know if you see any more problems of any kind - feedback and reports always welcome

I released [0.5.1](https://hackage.haskell.org/package/stack-clean-old-0.5.1) now to Hackage

Thank you for reporting this - kinda surprised noone did before.

**Comment by 0rphee** [comment]
I'm surprised too lol. 

Thanks for the quick fix :)
